### PR TITLE
Improve performance of Sort for the common single batch use case

### LINKED ISSUE #10570: [FEA] See if we can optimize sort for a single batch
**Is your feature request related to a problem? Please describe.**
In many cases a task is going to have a single batch to process. When we sort right now, we will sort the input batch, slice it into smaller batches, and then start to merge those batches together.  But most of that ends up being a noop if there is only a single batch to sort.

I would like to propose that we modify GpuOutOfCoreSortIterator so that it takes the first batch and just makes it spillable. It then asks for a second batch. Only if it gets a second batch does it go through the complicated sort/merge process. If there is only one batch we sort it and go on. It is a little more complicated than that because of retry and splits with the retry, but I think it is totally doable.

### PR DESCRIPTION
**Pull request by revans2**
This fixes #10570

The performance speed up is not that huge, but it is there.

I ran
```
spark.time(spark.range(0, 100000000L, 1, 12).selectExpr("id as oc2", "id DIV 10 as oc3", "CAST(id * 10 AS STRING) as oc", "CAST(id % 4 AS STRING) as pc", "id").selectExpr("*", "row_number() over (PARTITION BY pc ORDER BY oc, oc2, oc3) as rn", "max(id) OVER (PARTITION BY pc ORDER BY oc, oc2, oc3 ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) as m").orderBy(desc("oc")).show())
```

Both with this patch and without it. I captured the metric for sort time and the total run time of the query.

With this patch, on my desktop, the median run time was 6968 ms, and from the Spark UI the median sort time was 5.5 seconds.
Without this patch the runtime was 7051 and the sort time was 5.9 seconds.  That saves 83 ms (about 1% which is not really that huge), but the sort time metric showed about 0.4 seconds of savings or about 6% less time, which is a lot better.

### CODE REVIEW

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala [hunk 1]
```diff
@@ -178,6 +178,45 @@ case class GpuSortEachBatchIterator(
...
+      sortTime: GpuMetric = NoopMetric): Iterator[SpillableColumnarBatch] = {
+    val spillableIter = iter.flatMap { cb =>
+        // Filter out empty batches and make them spillable
+        if (cb.numRows() > 0) {
+          Some(SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY))
+        } else {
+          cb.close()
+          None
+        }
+    }
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:200]: If single batch is the common case it does not seem to matter but if we needed to save intermediate Option generation we could use an explicit PartialFunction instance:

```suggestion
    val spillableIter = iter.collect {
      // Filter out empty batches and make them spillable
      new PartialFunction[ColumnarBatch, SpillableColumnarBatch] {
        override def isDefinedAt(cb: ColumnarBatch): Boolean = if (cb.numRows() > 0) {
          true
        } else {
          cb.close()
          false
        }
        override def apply(cb: ColumnarBatch): SpillableColumnarBatch =
          SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
      }
    }
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala [hunk 2]
```diff
@@ -178,6 +178,45 @@ case class GpuSortEachBatchIterator(
...
+      iter: Iterator[ColumnarBatch],
+      sorter: GpuSorter,
+      opTime: GpuMetric = NoopMetric,
+      sortTime: GpuMetric = NoopMetric): Iterator[SpillableColumnarBatch] = {
+    val spillableIter = iter.flatMap { cb =>
+        // Filter out empty batches and make them spillable
+        if (cb.numRows() > 0) {
+          Some(SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY))
+        } else {
+          cb.close()
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:197]: What if this throws?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala [hunk 3]
```diff
@@ -398,45 +443,28 @@ case class GpuOutOfCoreSortIterator(
...
-        }
-      }
+  /**
+   * First pass through the data. Conceptually we are going to read in all of the batches, that are
+   * already sorted and split them up into smaller chunks for later merge sorting. But we are
+   * only going to do that if we have more than one batch to sort.
+   */
+  private final def firstPassReadBatches(scb: SpillableColumnarBatch): Unit = {
+    splitOneSortedBatch(scb)
+    while(alreadySortedIter.hasNext) {
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:466]: nit: space 
```suggestion
    while (alreadySortedIter.hasNext) {
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala (hunk 2 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:197]: When someone calls next the exception is thrown and they would need to handle it.

https://github.com/scala/scala/blob/1bdf362b0d8344927460116380b67fcaa766435b/src/library/scala/collection/Iterator.scala#L587-L620

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala (hunk 1 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:200]: I'll keep it in mind, but I'm not sure it matters that much here.

**Comment by revans2** [comment]
build

**Review by jlowe — APPROVED** [review]

**Review by gerashegalov — APPROVED** [review]

**Pull request merged**